Add basic text selection to editable text.

mpcomplete opened the pull request:
Only partially works. Editing the selected text doesn't work, which
probably will require engine changes. Currently only draws the selected
text and allows you to manipulate the selection with draggable selection
handles.

abarth commented on packages/flutter/lib/src/rendering/editable_line.dart:
```diff
@@ -111,6 +116,18 @@ class RenderEditableLine extends RenderBox {
+    Offset offset = _paintOffset + new Offset(0.0, -_kCaretHeightOffset);
+    Point left = selectionRects.first.toRect().bottomLeft + offset;
+    Point right = selectionRects.last.toRect().bottomRight + offset;
+    return [localToGlobal(left), localToGlobal(right)];
```
This isn't quite right for TRL.  You need to look at the textDirection of each rect to decide whether to take the left or right side.  For the first point, you want the leading side of the first rect (i.e., left for LTR and right for RTL).  For the second point, you want the trailing side of the last rect (i.e., right for LTR and left for RTL).

Also, just returning two points isn't always sufficient.  Some visual styles need to know the text direction at those points.  For example, if the selection handles aren't symmetric, they might want to orient them depending on the text direction at that point.

abarth commented on packages/flutter/lib/src/rendering/editable_line.dart:
```diff
@@ -177,14 +198,41 @@ class RenderEditableLine extends RenderBox {
+    int start = position.offset - 1;
+    while (start >= 0 && str[start] != ' ')
+      --start;
+    ++start;
```
Yeah, this only works in English (and not even really then).  You should ask the engine to compute the word boundaries for you.  It's a super complex process in some languages.  In C++, there's a WordIterator that knows how to do it.

abarth commented on packages/flutter/lib/src/widgets/editable.dart:
```diff
@@ -283,13 +288,41 @@ class RawInputLineState extends ScrollableState<RawInputLine> {
     if (config.onChanged != null)
       config.onChanged(_keyboardClient.inputValue.copyWith(selection: selection));
+
+    if (selection.baseOffset != selection.extentOffset)
```
`if (!selection.isCollapsed)`

abarth commented on packages/flutter/lib/src/widgets/editable.dart:
```diff
@@ -283,13 +288,41 @@ class RawInputLineState extends ScrollableState<RawInputLine> {
+    }
+  }
+
+  void _hideSelectionHandles() {
```
We also need to call this in `dispose`

abarth commented on packages/flutter/lib/src/widgets/editable.dart:
```diff
@@ -430,3 +463,153 @@ class _EditableLineWidget extends LeafRenderObjectWidget {
   }
 }
+
+enum _TextSelectionHandleSide { left, right }
```
Probably should be `start` and `end`.  Otherwise you're going to end up with an LTR bias.

abarth commented on packages/flutter/lib/src/widgets/editable.dart:
```diff
@@ -430,3 +463,153 @@ class _EditableLineWidget extends LeafRenderObjectWidget {
+  @override
+  void initState() {
+    super.initState();
+    selection = config.selection;
```
There shouldn't be any need to keep a separate reference to `selection`.  You can always get it from your `config`.

abarth commented on packages/flutter/lib/src/widgets/editable.dart:
```diff
@@ -430,3 +463,153 @@ class _EditableLineWidget extends LeafRenderObjectWidget {
+
+  @override
+  void didUpdateConfig(_TextSelectionHandleOverlay oldConfig) {
+    setState(() {
```
No need to call `setState` during `didUpdateConfig`.  You'll always get a chance to build after `didUpdateConfig` is called.

abarth commented on packages/flutter/lib/src/widgets/editable.dart:
```diff
@@ -430,3 +463,153 @@ class _EditableLineWidget extends LeafRenderObjectWidget {
+  void didUpdateConfig(_TextSelectionHandleOverlay oldConfig) {
+    setState(() {
+      selection = config.selection;
+      endpoints = config.renderObject.getEndpointsForSelection(selection);
```
Why not call this during build?  There doesn't seem to be any advantage in caching the information since you read it both in `initState` and in `didUpdateConfig`.

abarth commented on packages/flutter/lib/src/widgets/editable.dart:
```diff
@@ -430,3 +463,153 @@ class _EditableLineWidget extends LeafRenderObjectWidget {
+  }
+
+  Point _dragPosition;
+  void _dragStart(Point position, _TextSelectionHandleSide side) {
```
We usually call these functions `_handleDrageStart`.  We'd use `_dragStart` if this actually started the drag.

abarth commented on packages/flutter/lib/src/widgets/editable.dart:
```diff
@@ -430,3 +463,153 @@ class _EditableLineWidget extends LeafRenderObjectWidget {
+  }
+
+  void _dragUpdate(double delta, _TextSelectionHandleSide side) {
+    _dragPosition = _dragPosition + new Offset(delta, 0.0);
```
`+=`

abarth commented on packages/flutter/lib/src/widgets/editable.dart:
```diff
@@ -430,3 +463,153 @@ class _EditableLineWidget extends LeafRenderObjectWidget {
+
+    setState(() {
+      selection = newSelection;
+      endpoints = config.renderObject.getEndpointsForSelection(selection);
```
This isn't correct.  You're just going to blow way these fields when your parent rebuilds anyway.  Instead, you should just call onSelectionHandleChanged and rely on your parent rebuilding your with the new values.

mpcomplete replied:
I did it this way because I didn't think I could rebuild the OverlayEntry without removing it and re-adding - but now I see it has a markNeedsRebuild function. However, that will just call the builder function with the same values I originally passed in. Is the correct pattern to pull the builder out into a method, so that it can build the new _TextSelectionHandleOverlay with updated state that is cached on the calling class?

abarth replied:
Yes, that's fine.  If you leave the builder in a closure, you can still see the updated values for the member variables in the calling class.  It's probably cleaner to make the builder a member function though.

abarth commented on packages/flutter/lib/src/widgets/editable.dart:
```diff
@@ -430,3 +463,153 @@ class _EditableLineWidget extends LeafRenderObjectWidget {
+
+  @override
+  Widget build(BuildContext context) {
+    return new Stack(
```
Rather than creating a stack here, we should probably have two entries in the overlay, one for each of the selection handles.  When they move, they can call back into to the RawInputLineState which can merge the data, update the selection, and mark the entries as needing to rebuild.  Each selection handle can then operate independently.

The way you've got this set up currently doesn't handle multitouch properly (e.g., if one finger is dragging one handle at the same time as another finger is dragging another handle).  They'll fight over the _dragPosition.  If you set this up as two separate widgets with two separate overlay entries, you'll get multitouch correctness for free.

mpcomplete replied:
I've switched to this way and it works well, but I now realize I missed something. There's another text selection handle for when the TextSelection is collapsed. It might be nice to have a single overlay handle the TextSelection, and know whether to render it as either 2 handles or a single collapsed handle. Or maybe I should pull this out to a wrapper class that manages the selection... WDYT?

abarth replied:
For the collapsed case, you could also just make one of the handles be in charge (e.g., the base handle) and have the other handle be invisible (e.g., the extent handle).

abarth commented on packages/flutter/lib/src/widgets/editable.dart:
```diff
@@ -430,3 +463,153 @@ class _EditableLineWidget extends LeafRenderObjectWidget {
+
+  // Returns a rect of the given size at the corner of the selection range.
+  Rect _getCornerRect(double size) {
+    if (side == _TextSelectionHandleSide.left)
```
Please use a switch for branching over enums.

abarth commented on packages/flutter/lib/src/widgets/editable.dart:
```diff
@@ -430,3 +463,153 @@ class _EditableLineWidget extends LeafRenderObjectWidget {
+    if (side == _TextSelectionHandleSide.left)
+      return new Rect.fromLTWH(position.x - size, position.y, size, size);
+    else
+      return new Rect.fromLTWH(position.x, position.y, size, size);
```
Here's where you need to know whether the directionality of the selection handle.  It's very possible to have two handles pointed in the same direction if you combine a selection that starts in LTR text and ends in RTL text.

abarth commented on packages/flutter/lib/src/widgets/editable.dart:
```diff
@@ -430,3 +463,153 @@ class _EditableLineWidget extends LeafRenderObjectWidget {
+  void paint(Canvas canvas, Size size) {
+    // TODO(mpcomplete): pick a better color. Use an accent color based on the
+    // selection color?
+    Paint paint = new Paint()..color = new Color(0xff0000ff);
```
This should be based on the Theme.  The color probably needs to be given this to object from its widget, which can query the Theme for the textSelectionColor

abarth commented on packages/flutter/lib/src/widgets/editable.dart:
```diff
@@ -430,3 +463,153 @@ class _EditableLineWidget extends LeafRenderObjectWidget {
+    Rect rect = _getCornerRect(_kSelectionHandleRadius);
+    canvas.drawCircle(
+      (side == _TextSelectionHandleSide.left) ? rect.bottomLeft : rect.bottomRight,
+      _kSelectionHandleRadius, paint);
```
These are material-design selection handles.  Other design languages will want to use different styles of selection handles.  We'll need to factor the code such that `Input` (which is in the material layer) decides how the selection handles should appear.

abarth commented on packages/flutter/lib/src/widgets/editable.dart:
```diff
@@ -430,3 +463,153 @@ class _EditableLineWidget extends LeafRenderObjectWidget {
+  @override
+  bool shouldRepaint(_TextSelectionHandlePainter oldPainter) {
+    return (position != oldPainter.position ||
+            side != oldPainter.side);
```
No need for `(` or `)`

abarth commented on packages/flutter/lib/src/widgets/editable.dart:
```diff
@@ -430,3 +463,153 @@ class _EditableLineWidget extends LeafRenderObjectWidget {
+          onHorizontalDragStart: (Point p) { _dragStart(p, _TextSelectionHandleSide.left); },
+          onHorizontalDragUpdate: (double d) { _dragUpdate(d, _TextSelectionHandleSide.left); },
+          child: new CustomPaint(
+            foregroundPainter: new _TextSelectionHandlePainter(
```
Why a foreground painter and not just a regular painter?  This widget doesn't have any children.

Also, what determines the size of the CustomPaint widget?

mpcomplete replied:
The handles themselves are painted on top of stuff, so I assumed they should be drawn with a foreground painter. The size is calculated by the painter, which has its custom hitTest method. Am I misunderstanding the CustomPaint concept?

abarth commented on packages/flutter/lib/src/widgets/editable.dart:
```diff
@@ -430,3 +463,153 @@ class _EditableLineWidget extends LeafRenderObjectWidget {
+  }
+
+  @override
+  bool hitTest(Point position) {
```
You shouldn't need this method.  The default logic in CustomPaint should be sufficient.

mpcomplete replied:
How does the CustomPaint know its position/size?

abarth replied:
The widget that `Input` passes into the `RawInputLine` could have some preferred size (e.g., by including a `SizedBox` in its build).  The harder part is knowing where to place that box relative to the selection endpoint...  I guess it's reasonable to use a full-screen custom painter that only draws (and hit tests to) a small part of the screen.

abarth commented:
w.r.t. letting `Input` control the visual appearance of the selection handles, I'd probably let `Input` supply two widgets to `RawInputLine`, one for each direction of selection handle.  Most of the time, `RawInputLine` wouldn't include these in its build but when it could use them when it wanted to draw selection handles.  If it needs two "leftward" handles (e.g., because the selection starts in LTR text and ends in RTL text), then it can include the "leftward" selection handle twice.  (That's a perfectly find thing to do.)  Then the custom painting logic can be controlled at the material layer but the gesture detection and selection adjustment can be handled by `RawInputLine`.

mpcomplete commented:
Thanks for the detailed feedback! I'm new to text handling, so I'm not surprised I missed all these things. I added a couple questions to your comments while I work on fixing the rest.

abarth commented:
Text selection is tricky because of RTL.  I'm happy to jump on VC and we can draw some diagrams on a whiteboard if that's helpful.

Hixie commented:
For the record it's possible that whatever approach we take here we'll have to revisit it soonish to make this work on Android according to platform conventions, which is to have the handles be able to go outside the application boundaries and into system-drawn UI (even into the soft buttons at the bottom of the screen!).

On the other hand, on iOS, the platform convention appears to be to have the handles clipped to the text widget boundaries, so not even in the system overlay. (It would actually be trivial to do this by just wrapping the raw input line control with an Overlay of its own, since then the input control would find _that_ Overlay when doing `Overlay.of()`. So long as all the calculations are correctly doing local-to-global-to-local conversions, it would Just Work.)

mpcomplete commented:
OK, I made a bunch of changes and it's ready for another look. Lots has changed from the first patch - sorry about that. Major points:
- I pulled the most of the handle management out of RawInputLineState into a separate class for cleanliness.
- I moved the actual handle drawing to material/input.dart, passing a builder function down to the lower layer.
- I'm using SizedBox and Positioned to place the handles, so I no longer need a custom hit tester.
- We use a single-point handle for collapsed selections. The behavior matches Android as best I could. (Though for now I'm drawing it with a 'left-side' handle.)
- LTR/RTL works much better now.

I still need to fix word selection. The placeholder works OK for testing, so I'd like to do that in a separate patch.

abarth commented on packages/flutter/lib/src/rendering/editable_line.dart:
```diff
@@ -17,6 +16,16 @@ const double _kCaretWidth = 1.0; // pixels
+class TextSelectionPoint {
+  TextSelectionPoint(this.point, this.direction);
+  Point point;
+  TextDirection direction;
```
final

abarth commented on packages/flutter/lib/src/rendering/editable_line.dart:
```diff
@@ -17,6 +16,16 @@ const double _kCaretWidth = 1.0; // pixels
+/// text direction at that point.
+class TextSelectionPoint {
+  TextSelectionPoint(this.point, this.direction);
+  Point point;
```
final

abarth commented on packages/flutter/lib/src/rendering/editable_line.dart:
```diff
@@ -111,6 +122,31 @@ class RenderEditableLine extends RenderBox {
   }
 
+  List<TextSelectionPoint> getEndpointsForSelection(TextSelection selection) {
+    _textPainter.layout(); // TODO(mpcomplete): is this hacky?
```
Yeah, we need to re-think how we handle the text painter's layout more generally.  Not something we need to solve in this patch.

abarth commented on packages/flutter/lib/src/rendering/editable_line.dart:
```diff
@@ -111,6 +122,31 @@ class RenderEditableLine extends RenderBox {
+      // TODO(mpcomplete): This doesn't work well at an RTL/LTR boundary.
+      Offset caretOffset = _textPainter.getOffsetForCaret(_selection.extent, _caretPrototype);
+      Point start = new Point(caretOffset.dx, _contentSize.height) + offset;
+      return [new TextSelectionPoint(localToGlobal(start), TextDirection.ltr)];
```
Why ltr?

abarth commented on packages/flutter/lib/src/rendering/editable_line.dart:
```diff
@@ -111,6 +122,31 @@ class RenderEditableLine extends RenderBox {
+  }
+
+  TextPosition getPositionForPoint(Point global) {
+    return _textPainter.getPositionForOffset(globalToLocal(global).toOffset());
```
We don't need to take the `_paintOffset` into account here?

abarth commented on packages/flutter/lib/src/widgets/editable.dart:
```diff
@@ -318,6 +348,9 @@ class RawInputLineState extends ScrollableState<RawInputLine> {
       _keyboardHandle.release();
     if (_cursorTimer != null)
       _stopCursorTimer();
+    Timer.run(() { // can't hide while disposing, since it triggers a rebuild
```
scheduleMicrotask rather than Timer.  You want this to run before we process any more events.

abarth commented on packages/flutter/lib/src/widgets/editable.dart:
```diff
@@ -341,6 +374,13 @@ class RawInputLineState extends ScrollableState<RawInputLine> {
       _stopCursorTimer();
 
+    if (_selectionHandles != null && !focused) {
+      Timer.run(() { // can't hide while building, since it triggers a rebuild
```
scheduleMicrotask again for the same reason.  For example, you wouldn't want user input to sneak in during the async gap because the selection handles would probably get mad if the user interacted with them in that state.

abarth commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -0,0 +1,223 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
```
2016

abarth commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -0,0 +1,223 @@
+    <_TextSelectionHandlePosition>[
+      _TextSelectionHandlePosition.start,
+      _TextSelectionHandlePosition.end,
+    ];
```
Why not just use _TextSelectionHandlePosition.values ?

mpcomplete replied:
Huh.. yeah, I don't need an index at all. I swear it mattered at one point. :)

FIxed

abarth commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -0,0 +1,223 @@
+  final RenderEditableLine renderObject;
+  final ValueChanged<TextSelection> onSelectionHandleChanged;
+  final TextSelectionHandleBuilder builder;
+  TextSelection selection;
```
Do you mean for the setter for this value to be public?  I would have expected mutable values to be private.

abarth commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -0,0 +1,223 @@
+  /// Shows the handles by inserting them into the [context]'s overlay.
+  void show(BuildContext context, { Widget debugRequiredFor }) {
+    assert(_handles == null);
+    _handles = [
```
Add `<OverlayEntry>` for more type checking.

abarth commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -0,0 +1,223 @@
+
+  Widget _buildOverlay(BuildContext context, int which) {
+    if (selection.isCollapsed && which == 1)
+      return new OffStage(child: null);  // hide the second handle when collapsed
```
We usually just return `new Container()` when we want an blank widget.

abarth commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -0,0 +1,223 @@
+
+    TextSelection newSelection;
+    switch (config.position) {
+      case _TextSelectionHandlePosition.start:
```
I'd probably use base/extent for these enum values because they map down to baseOffset and extentOffset, respectively, in TextSelection, but what you've got is also fine.

mpcomplete replied:
I went back and forth between the two. I settled on start/end because it conveys the intention that `start` will always be logically before `end` (in terms of offset into the text).

abarth commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -0,0 +1,223 @@
+            child: _buildChild(context, type)
+          )
+        ]
+      )
```
No need for a `Stack` here.  The Overlay already has a stack, so you can use a `Positioned` directly.  You just need to put it outside the GestureDetector so that it's the first widget returned by this build function.

mpcomplete replied:
That seems to break hit detection. Maybe GestureDetector needs the positioning info?

mpcomplete replied:
Though it's only broken for the `start` handle because it has a FractionalTranslate lower down the hierarchy. If I remove the FractionalTranslate, it works again. Could I be hitting a bug?

abarth replied:
It's not a bug.  The fractional translation is moving the child outside of the bounds of its parent.  The hit test walk is pruned at each stage by the bounds of the render object, which means the hit test is pruned before it gets to the fractional translation.

The way I'd recommend doing this is to pass the position down to the child builder and have the child use that information to position itself relative to the selection point.  That way it can place itself wherever it likes.  Because you want to wrap the child in a gesture detector, the child will need to use something like a Transform to move itself into position.

mpcomplete replied:
So you're saying don't use a Positioned at all, but a Transform instead?

abarth replied:
Yeah, I'd give that a try.

mpcomplete replied:
Is there a reason to prefer Transform to Stack+Positioned? A SizedBox inside a Transform doesn't work right (I assume the Transform is applying to the size as well). I could put the Transform inside the SizedBox, but that would require (as you say) passing the position to the handle builder to position itself. I like the way the overlay is responsible for positioning each handle correctly, and the handle builder simply returns a widget to display at that position (with possible relative offsets).

abarth replied:
Stack+Positioned is just a bit of a roundabout way to position something.  There's nothing wrong with it particularly.

abarth commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -0,0 +1,223 @@
+    TextSelectionPoint endpoint,
+    TextSelectionHandleType ltrType,
+    TextSelectionHandleType rtlType
+  ) {
```
This might be slightly more readable at the call site if you used named arguments here.

abarth commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -0,0 +1,223 @@
+    switch (type) {
+      case TextSelectionHandleType.left:
+        // Shift the child left by 100% of its width, so its top-right corner
+        // touches the selection endpoint.
```
This is true for material-design selection handles but isn't true for iOS selection handles.  On iOS, the selection handles are centered (horizontally) at the point of selection.  Why not push this work into the `config.builder` function?

abarth commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -0,0 +1,223 @@
+          child: child
+        );
+        break;
+      default: break;
```
It's probably better to list out the other enum value in an explicit case statement.

Also it's slightly clear if you return direction from inside the switch instead of breaking out to return.

abarth commented:
LGTM

Please add some tests before landing.  If you take a look at some of the existing tests, you can see how to mock out the keyboard service.  Probably makes sense to write a test that shows you can long-press to create a selection and then you can drag the two selection handles to change the selection.

mpcomplete commented:
Added some tests. Waiting on travis.

I also added a new Color to Theme, `textSelectionHandleColor`, for the handle color. It's a slightly darker version of the selected text color.

mpcomplete committed:
Add basic text selection to editable text.

Only partially works. Editing the selected text doesn't work very well,
which probably will require engine changes. Currently only draws the
selected text and allows you to manipulate the selection with draggable
selection handles.

sethladd commented:
Thanks! Curious, did you test this on iOS?

mpcomplete commented:
Arg, no! I will do so.
